subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -110,36 +136,106 @@ private boolean experimentalCompilationAvoidanceEnabled() {
     }
 
     @Override
-    @TaskAction
     protected void compile() {
+        compile(null);
+    }
+
+    @TaskAction
+    protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
+
+        this.sourceClassesMappingFile = new File(getTemporaryDir(), "source-classes-mapping.txt");
```

Comment:
`JavaCompile` doesn't need to handle this. It has no state file - all it needs to do after the compilation is storing the whole classpath and output directory into cache, see `IncrementalResultStoringCompiler`.

Reply:
But I still think this file is necessary - we need to keep the source file-class relationship in case that the next build is incremental compilation.

Reply:
Is it correct that by "incremental compilation" you mean "incremental compilation enabled via option.incremental"?

That is not the problem I am talking about. If you load some previous output from the cache, then the mapping file and the actual sources classes are out of sync, since the mapping file is from the build which actually executed the task action, while the classes have been loaded from the cache. Doing an incremental compilation (as in "only compile a few files") in the next execution may lead incorrect results. Imagine a `.groovy` file now generates a completely different set of classes.

Reply:
We never write the mapping file when the task is up-to-date or from the cache. So we could share the mapping file via the build cache - that seems to be harder though given that we currently rely quite heavily on absolute paths.

Note that we may have a problem with up-to-date builds already: if a source file moves from one source root to another, then the mapping file is also incorrect (wrong source file key), while the task is up-to-date.

Reply:
I made it `@LocalState` and added a test: https://github.com/gradle/gradle/blob/41adfd42e7cf5eaa9f06f80cf9d6812f8dbca5d6/subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/GroovySourceIncrementalCompilationIntegrationTest.groovy#L34